linera-execution/src/runtime.rs

```diff
@@ -1232,16 +1180,28 @@ impl ContractRuntime for ContractSyncRuntimeHandle {
 
     fn send_message(&mut self, message: SendMessageRequest<Vec<u8>>) -> Result<(), ExecutionError> {
         let mut this = self.inner();
-        let application = this.current_application_mut();
+        let application = this.current_application();
+        let authenticated_signer = application.signer;
+        let application_id = application.id;
 
-        application.outcome.messages.push(message.into());
+        let message =
+            RawOutgoingMessage::from(message).into_priced(&this.resource_controller.policy)?;
+        this.resource_controller.track_grant(message.grant)?;
+
+        let outcome = RawExecutionOutcome {
+            authenticated_signer,
+            refund_grant_to: this.refund_grant_to,
+            messages: vec![message],
+        };
+        this.transaction_tracker
+            .add_user_outcome(application_id, outcome)?;
```

Comment:
(Not in this PR, but: This could just become `add_message`, and take an `OutgoingMessage`, which whould have to be moved to `linera-execution`, could it? We don't need `RawExecutionOutcome` at all, and after this PR it will always contain only a single message anyway.)